sale_timesheet_line_exclude/models/account_analytic_line.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2018-2019 Brainbean Apps
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import fields, models, api, _
+from odoo.exceptions import UserError
+
+
+class AccountAnalyticLine(models.Model):
+    _inherit = 'account.analytic.line'
+
+    exclude_from_sale_order = fields.Boolean(
+        string='Exclude from Sale Order',
+        help=(
+            'Checking this would exclude this timesheet entry from Sale Order'
+        ),
+    )
+
+    @api.onchange('task_id', 'employee_id')
+    def _onchange_task_id_employee_id(self):
+        if self.project_id:  # timesheet only
+            self.so_line = self._timesheet_get_sale_line()
+
+    @api.onchange('exclude_from_sale_order')
+    def _onchange_exclude_from_sale_order(self):
+        if self.project_id:  # timesheet only
+            self.so_line = self._timesheet_get_sale_line()
+
+    @api.multi
+    def _timesheet_get_sale_line(self):
+        self.ensure_one()
+        return self._timesheet_determine_sale_line(
+            **self._timesheet_determine_sale_line_arguments()
+        ) if not self.exclude_from_sale_order else False
+
+    @api.model
+    def _timesheet_should_evaluate_so_line(self, values, check):
+        return check([field_name in values for field_name in [
+            'task_id',
+            'employee_id',
+            'exclude_from_sale_order',
+        ]])
+
+    @api.multi
+    def _timesheet_determine_sale_line_arguments(self, values=None):
+        return {
+            'task': (
+                self.env['project.task'].sudo().browse(values['task_id'])
+            ) if values and 'task_id' in values else self.task_id,
+            'employee': (
+                self.env['hr.employee'].sudo().browse(values['employee_id'])
+            ) if values and 'employee_id' in values else self.employee_id,
+        }
+
+    @api.multi
+    @api.depends('exclude_from_sale_order')
+    def _compute_timesheet_invoice_type(self):
+        result = super()._compute_timesheet_invoice_type()
+
+        for timesheet in self:
+            # Skip non-timesheet entries
+            if not timesheet.project_id:
+                continue
+
```

Comment:
You know that I don't like too much these useless empty lines around the code.

Reply:
Removed

Reply:
All of them?

Reply:
Most of them

Reply:
OK, I'm content with that